### PR TITLE
Friendship v1

### PR DESCRIPTION
**Pull request by Morrism1**
# Milestone 5: Friendship V1

* Created friendship Invitation model
* Added associations between user and friendship invitation models
* Added new routes to send a friendship invitation features
* Added button link to request friendship in users views
* Added links to accept or decline the friendship in friendship's index page
* Added integration and unit tests with RSpec, capybara, and shoulda for all the features and methods
* You can run rails DB:SEED for a quick 15 users creation with faker gem 👌

### CODE REVIEW

File: app/helpers/friendship_invitations_helper.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module FriendshipInvitationsHelper
+  def friend_request?(user)
+    current_user.friendship_invitations.find_by(friend_id: user.id).nil?
```
Comment (tirthajyoti-ghosh) [inline app/helpers/friendship_invitations_helper.rb]: `find_by` method queries the database to find a row with given constraints. That means it's processing data (application logic), which doesn't belong in "helpers". Helpers are reserved for putting the view logic, not the application logic.

Application logic belongs in the `model`. Please move this logic to the model and use it in controllers. Everything connected with data processing should be encapsulated in models and used in controllers.

File: app/helpers/friendship_invitations_helper.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  def friend_requests
+    friends.map { |friend| friend.user unless friend.status }.compact
+  end
+
+  def confirm_request(user)
+    friend = friends.find { |invite| invite.user == user }
+    friend.status = true
+    friend.save
+  end
```
Comment (tirthajyoti-ghosh) [inline app/helpers/friendship_invitations_helper.rb]: Please do the same thing for this method as well. Please move this logic to the model and use it in controllers.

_**Remember** to update your tests!_

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -19,6 +19,9 @@
         </div>
       </div>
       <% if current_user %>
+        <div class="menu-item">
+          <%= menu_link_to 'Friend Requests', user_friendships_path(current_user.id) %>
+        </div>
         <%= link_to 'Sign out', destroy_user_session_path, method: :delete %>
       <% else %>
```
Comment (tirthajyoti-ghosh) [inline app/views/layouts/application.html.erb]: Please remove all logic from views. There should be absolutely no logic in views. Instead, outsource all your logic to helpers. There are two more `if` statements below this. Remove them as well.

**Resources**
- [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
- [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
- [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

**Review by tirthajyoti-ghosh — CHANGES_REQUESTED** [review]
# Changes required 🛑 

## What a remarkable project you have built 💯! This is impressive 🤩👌🧨!

### But in order to get it approved, we need to work on some changes. Let's see...

- You have missed a couple of tests. Please add tests for the remaining models:
  - [x] `comment`
  - [x] `like`
  - [x] `post`

  Please add controller tests. There are currently no controller tests.

- Right now, the timeline displays posts from all users. As per the requirements of the project, the timeline should display posts from the logged-in user and the user's friends (users who have accepted the logged-in user's friend request) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

  - [x] Please implement this functionality.

Your project looks absolutely fantastic. Just make these changes and apply for another code review.

Happy coding 👨‍💻 

*Need help? Shoot me a message on Slack - [@Ghosh](https://app.slack.com/client/T47CT8XPG/DRDM6LAQN/user_profile/URB1MHXK6)*
<hr>

**Tirthajyoti Ghosh**

[GitHub](https://github.com/tirthajyoti-ghosh) | [LinkedIn](https://www.linkedin.com/in/tirthajyoti-ghosh/) | [Twitter](https://twitter.com/terrific_ghosh)

**Review by trekab — APPROVED** [review]
### Approved! :tada: 
Well done @Morrism1! Nice job implementing the suggested changes. You may proceed to the next milestone.

_Happy coding!_ :wink: 
__Treasure Kabareebe__
[GitHub](https://github.com/trekab) | [LinkedIn](https://www.linkedin.com/in/treasure-kabareebe/) | [Twitter](https://twitter.com/TKabareebe)

**Pull request merged**